Assets/Mirror/Runtime/SyncDictionary.cs

```diff
@@ -0,0 +1,297 @@
+                        SerializeItem(writer, change.item);
+                        break;
+                    case Operation.OP_CLEAR:
+                        break;
+                }
+            }
+        }
+
+        public void OnDeserializeAll(NetworkReader reader)
+        {
+            // This list can now only be modified by synchronization
+            IsReadOnly = true;
+
+            // if init,  write the full list content
+            int count = (int)reader.ReadPackedUInt32();
+
+            m_Objects.Clear();
+            Changes.Clear();
+
+            for (int i = 0; i < count; i++)
+            {
+                B key = DeserializeKey(reader);
+                T obj = DeserializeItem(reader);
+                m_Objects.Add(key, obj);
+            }
+
+            // We will need to skip all these changes
+            // the next time the list is synchronized
+            // because they have already been applied
+            changesAhead = (int)reader.ReadPackedUInt32();
+        }
+
+        public void OnDeserializeDelta(NetworkReader reader)
+        {
+            // This list can now only be modified by synchronization
+            IsReadOnly = true;
+
+            int changesCount = (int)reader.ReadPackedUInt32();
+
+            for (int i = 0; i < changesCount; i++)
+            {
+                Operation operation = (Operation)reader.ReadByte();
+
+                // apply the operation only if it is a new change
+                // that we have not applied yet
+                bool apply = changesAhead == 0;
+                B key = default;
+                T item = default;
+
+                switch (operation)
+                {
+                    case Operation.OP_ADD:
+                    case Operation.OP_SET:
+                    case Operation.OP_DIRTY:
+                        key = DeserializeKey(reader);
+                        item = DeserializeItem(reader);
+                        if (apply)
+                        {
+                            m_Objects[key] = item;
+                        }
+                        break;
+
+                    case Operation.OP_CLEAR:
+                        if (apply)
+                        {
+                            m_Objects.Clear();
+                        }
+                        break;
+
+                    case Operation.OP_REMOVE:
+                        key = DeserializeKey(reader);
+                        item = DeserializeItem(reader);
+                        if (apply)
+                        {
+                            m_Objects.Remove(key);
+                        }
+                        break;
+                }
+
+                if (apply)
+                {
+                    Callback?.Invoke(operation, key, item);
+                }
+                // we just skipped this change
+                else
+                {
+                    changesAhead--;
```

Comment:
--changesAhead would be more obvious imho. changesAhead-- does one more thing. compiler probably ignores it in this case, but I still find it a bit strange to look at

Reply:
Compiler changes --i to i-- when it doesn't matter

Reply:
why would it change --i to i--? you probably mean the other way around?

Reply:
No, it changes to i--

Reply:
do you have a source for that?

Reply:
Put i++ and ++i in code, decompile it and see what it does

Reply:
I think you are confusing the two.
it's very unlikely that the IL code for
changesAhead-- would end up as '(changesAhead+=1)-1' - there is no reason for the compiler to add one extra unnecessary operation. you probably meant it the other way around, that it changes i-- to --i.

Reply:
https://stackoverflow.com/questions/6889647/is-there-any-performance-difference-with-i-vs-i-1-in-c IL code is basically i = i + 1

Reply:
@vis2k both compile to the exact same thing, if you want to enforce this I recommend putting it in contribution guidelines, since its a completely arbitrary decision. Although I see where you are coming from, that postincrement is theoretically more complicated because of the copy op, in practice there is no actual difference if the return value is not used. Since that is the case here, I side with Petris and katori, i++ is the C# standard. Everyone writes for loops this way.